vllm/engine/spec_dec.py

```diff
@@ -0,0 +1,138 @@
+        max_len = max([len(input_ids) for input_ids in input_ids_list])
+        input_ids_list = [_pad_left_to_max(input_ids, max_len, PAD_TOKEN_ID) for input_ids in input_ids_list]
+        return torch.tensor(input_ids_list, dtype=torch.long, device='cuda')
+    
+    # TODO: we need to align draft and target model's sampler
+    def _sample_method(self, logits):
+        temperature = 1.0
+        return torch.softmax(logits / temperature, dim=-1)
+    
+    def set_draft_tokens(self,
+                seq_group_list: List[SequenceGroupMetadata]) -> None:
+        logger.info(f"# of input request: {len(seq_group_list)}")
+        input_tensor = self._prepare_inputs(seq_group_list)
+        draft_logits, draft_distributions, draft_tokens = [], [], []
+        # recompute for now
+        attention_mask=(input_tensor != PAD_TOKEN_ID)
+        past_key_values = None
+        for i in range(self.propose_cnt):
+            with torch.no_grad():
+                outputs = self.draft_model(input_tensor,
+                                    past_key_values=past_key_values,
+                                    attention_mask=attention_mask,
+                                    use_cache=True)
+            
+            past_key_values = outputs.past_key_values
+            next_token_logits = outputs.logits[:, -1, :]
+            distribution = self._sample_method(next_token_logits)
+            attention_mask = torch.cat([attention_mask, torch.ones(input_tensor.shape[0], 1, device='cuda')], dim=1)
+            input_tensor = torch.multinomial(distribution, num_samples=1)
+            
+            draft_logits.append(next_token_logits)
+            draft_distributions.append(distribution)
+            draft_tokens.append(input_tensor)
+
+        for i, seq_group_metadata in enumerate(seq_group_list):
+            seq_id = next(iter(seq_group_metadata.seq_data))
+            seq = seq_group_metadata.seq_data[seq_id]
+            for j in range(self.propose_cnt):
+                draft_token = draft_tokens[j][i].item()
+                seq.draft_token_probs.append({draft_token:draft_distributions[j][i]})
+                seq.draft_token_ids.append(draft_token)
+            logger.info(f"Seq draft tokens: {seq.draft_token_ids}")
+            # logger.info(f"Seq draft prob: {seq.draft_token_probs}")
+        
+    def accept(self,
+               target_outputs: List[SamplerOutput]):
+        def extract_draft_prob_dis(sample: SequenceOutputs,
+                                   token_id: int,
+                                   index: int):
+            token_prob = sample.sd_draft_probs[index]
+            assert token_id in token_prob
+            return token_prob[token_id]
+        
+        def extract_target_prob_dis(seq_grou_output: SequenceGroupOutputs,
+                                    token_id: int):
+            # TODO: implement this
+            vocab_size = 50272
+            return torch.rand(1, vocab_size, device='cuda').squeeze(0)
```

Comment:
I would say it might be better to define these two functions out of this method, otherwise, the definition code will be run every time you call `accept`. vllm is a performance sensitive server so we might want to be carefully as we can.